### DISCUSSION TITLE
Create an I2C device with MicroPython (Pico Project)

### DISCUSSION THREAD

**Discussion by mchobby** [micropython/micropython#9791] [Using MicroPython]
Hi,
I would like to create a MicroPython Device that expose an I2C interface.
Creating an I2C client (to control an I2C device) does looks obvious to me.
However creating an I2C "server" (to being controlled over I2C) is not so obvious.
I did find any resource/thread/example on that subject.

Does  anyone have a suggestion?

**Comment by jimmo** [comment]
https://forum.micropython.org/viewtopic.php?t=9736#p54482

(Summary: i2c peripheral is supported by the hardware, but no MicroPython API for it currently... some thought needs to go into the design)

**Reply by mchobby** [reply]
This example  definitively a Winner, I successfully tested it with 2 Pico.

Here is my micropython test client side done with a second Pico on I2C(1) - GP6=sda, GP7=scl .
Don't forget to place pull-up resistor to 3.3V (10 KOhms) on SDA and SCL lines.
 
```
>>> from machine import I2C
>>> i2c = I2C( 1 )
>>> # Read one byte from register 0
>>> i2c.readfrom_mem( 0x41, 0, 1 )
b'\x00'

>>> # Write a value to a register and read it back!
>>> d = bytes([123])
>>> d
b'{'
>>> i2c.writeto_mem( 0x41, 0, d )
>>> i2c.readfrom_mem( 0x41, 0, 1 )
b'{'

>>> #read register1 to 4 to get 32 bits floating value
>>> buf4 = bytearray(4)
>>> i2c.readfrom_mem_into(0x41,1,buf4)
>>> buf4
bytearray(b'N\xec\xa3A')

>>> # Decoding the temperature
>>> import struct
>>> struct.unpack('f',buf4)
(20.49038,)
```

**Comment by Lobo-T** [comment]
This has been discussed on the Raspberry Pi forum.
Here is a solution by manipulating registers with mem32 statements:
https://forums.raspberrypi.com/viewtopic.php?p=2000723#p2000723
